File: android/src/main/java/expo/modules/motionblurview/ExpoMotionBlurViewModule.kt
package expo.modules.motionblurview

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.motionblurview.models.MotionBlurConfiguration
import expo.modules.motionblurview.views.MotionBlurView
import kotlin.math.max

class ExpoMotionBlurViewModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ExpoMotionBlurView")

    View(MotionBlurView::class) {
      val defaults = MotionBlurConfiguration.DEFAULT

      Prop("intensity") { view: MotionBlurView, value: Double? ->
        view.configuration = view.configuration.copy(intensity = max(0.0, value ?: defaults.intensity))
      }

      Prop("speedForMaxBlur") { view: MotionBlurView, value: Double? ->
        view.configuration = view.configuration.copy(speedForMaxBlur = max(1.0, value ?: defaults.speedForMaxBlur))
      }

      Prop("samples") { view: MotionBlurView, value: Int? ->
        val samples = (value ?: defaults.maxSampleCount).coerceIn(MotionBlurConfiguration.SAMPLE_COUNT_RANGE)
        view.configuration = view.configuration.copy(maxSampleCount = samples)
      }

      Prop("enabled") { view: MotionBlurView, value: Boolean? ->
        view.configuration = view.configuration.copy(isEnabled = value ?: defaults.isEnabled)
      }

      Prop("live") { _: MotionBlurView, _: Boolean? -> }

      OnViewDidUpdateProps { view: MotionBlurView ->
        view.propsDidUpdate()
      }
    }
  }
}

File: ios/Module/ExpoMotionBlurViewModule.swift
//
//  ExpoMotionBlurViewModule.swift
//  Pods
//
//  Created by rit3zh CX on 9/23/26.
//

import ExpoModulesCore

public class ExpoMotionBlurViewModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoMotionBlurView")

    OnCreate {
      DispatchQueue.main.async {
        MetalContext.shared?.prepare()
      }
    }

    View(MotionBlurView.self) {
      let defaults = MotionBlurConfiguration.default

      Prop("intensity", Double(defaults.intensity)) { (view: MotionBlurView, value: Double) in
        view.configuration.intensity = max(0, CGFloat(value))
      }

      Prop("speedForMaxBlur", Double(defaults.speedForMaxBlur)) { (view: MotionBlurView, value: Double) in
        view.configuration.speedForMaxBlur = max(1, CGFloat(value))
      }

      Prop("samples", defaults.maxSampleCount) { (view: MotionBlurView, value: Int) in
        view.configuration.maxSampleCount = min(max(value, MotionBlurConfiguration.sampleCountRange.lowerBound), MotionBlurConfiguration.sampleCountRange.upperBound)
      }

      Prop("enabled", defaults.isEnabled) { (view: MotionBlurView, value: Bool) in
        view.configuration.isEnabled = value
      }

      Prop("live", defaults.isLive) { (view: MotionBlurView, value: Bool) in
        view.configuration.isLive = value
      }

      OnViewDidUpdateProps { (view: MotionBlurView) in
        view.propsDidUpdate()
      }
    }
  }
}
